### ISSUE TITLE
flag lists for resolve filter are undocumented

### ISSUE THREAD

**Issue by purpleidea** [hughsie/PackageKit#113]
I'm trying to figure out the flags to pass to the resolve function.

https://github.com/hughsie/PackageKit/blob/master/src/org.freedesktop.PackageKit.Transaction.xml#L1091

It seems it takes these flags or filters (or some of them):

https://www.freedesktop.org/software/PackageKit/gtk-doc/introduction-ideas-filters.html

But it's unclear which ones, and how to use them. The DBUS api seems to take a 64bit uint, but how to build this up with the correct flags?

Docs and/or suggestions welcome :)

**Comment by ximion** [comment]
pk-bitfield.c might be interesting for you, e.g. https://github.com/hughsie/PackageKit/blob/master/lib/packagekit-glib2/pk-bitfield.c#L218

**Comment by purpleidea** [comment]
Still not particularly obvious to a non-c programmer using the dbus api. If this could be spelled out more explicitly (preferably in the docs) it would be appreciated. Cheers!

**Comment by purpleidea** [comment]
More digging found this:

https://github.com/hughsie/PackageKit/blob/master/lib/packagekit-glib2/pk-enum.c#L215

It looks to be the order the values go in... No idea how your API doesn't break if you figure out which value is which and someone changes the code. Pretty sure a string_to_flags method should probably be provided somewhere for dbus users.

**Comment by purpleidea** [comment]
I've put the following in golang:

```
const ( //static const PkEnumMatch enum_filter[]
	PK_FILTER_ENUM_UNKNOWN         uint64 = 1 << iota // "unknown"
	PK_FILTER_ENUM_NONE                               // "none"
	PK_FILTER_ENUM_INSTALLED                          // "installed"
	PK_FILTER_ENUM_NOT_INSTALLED                      // "~installed"
	PK_FILTER_ENUM_DEVELOPMENT                        // "devel"
	PK_FILTER_ENUM_NOT_DEVELOPMENT                    // "~devel"
	PK_FILTER_ENUM_GUI                                // "gui"
	PK_FILTER_ENUM_NOT_GUI                            // "~gui"
	PK_FILTER_ENUM_FREE                               // "free"
	PK_FILTER_ENUM_NOT_FREE                           // "~free"
	PK_FILTER_ENUM_VISIBLE                            // "visible"
	PK_FILTER_ENUM_NOT_VISIBLE                        // "~visible"
	PK_FILTER_ENUM_SUPPORTED                          // "supported"
	PK_FILTER_ENUM_NOT_SUPPORTED                      // "~supported"
	PK_FILTER_ENUM_BASENAME                           // "basename"
	PK_FILTER_ENUM_NOT_BASENAME                       // "~basename"
	PK_FILTER_ENUM_NEWEST                             // "newest"
	PK_FILTER_ENUM_NOT_NEWEST                         // "~newest"
	PK_FILTER_ENUM_ARCH                               // "arch"
	PK_FILTER_ENUM_NOT_ARCH                           // "~arch"
	PK_FILTER_ENUM_SOURCE                             // "source"
	PK_FILTER_ENUM_NOT_SOURCE                         // "~source"
	PK_FILTER_ENUM_COLLECTIONS                        // "collections"
	PK_FILTER_ENUM_NOT_COLLECTIONS                    // "~collections"
	PK_FILTER_ENUM_APPLICATION                        // "application"
	PK_FILTER_ENUM_NOT_APPLICATION                    // "~application"
	PK_FILTER_ENUM_DOWNLOADED                         // "downloaded"
	PK_FILTER_ENUM_NOT_DOWNLOADED                     // "~downloaded"
	//0, NULL}
)
```

This seems to work, until it doesn't. It's not clear if the parts that don't work are because if packagekit bugs, or that I got the "bitfield" wrong, or that I misunderstood how the library is supposed to work.

In this example, the first listed value gets a "1", the second gets a "2", the third gets a "4", next a "8" and so on... If this is correct/not correct, including what the initial value should be, please let me know.

Secondly, the PK_FILTER_ENUM_NOT_NEWEST filter for example, returns incorrect things. It returns:

```
gedit;2:3.18.3-1.fc23;x86_64;installed:updates
sl;5.02-2.fc23;x86_64;installed:@commandline
znc;1.6.2-1.fc23;x86_64;installed:@commandline
gedit;2:3.18.3-1.fc23;i686;updates
powertop;2.8-1.fc23;x86_64;updates
gedit;2:3.18.1-1.fc23;i686;fedora
gedit;2:3.18.1-1.fc23;x86_64;fedora
irssi;0.8.17-4.fc23;x86_64;fedora
powertop;2.7-6.fc23;x86_64;fedora
znc;1.6.1-1.fc23;x86_64;fedora
```

when you ask it for "powertop", "sl", "irssi", "gedit", "znc". Conversely, the PK_FILTER_ENUM_NEWEST filter DOES seem to return what you expect!

Is this another bug?

**Comment by hughsie** [comment]
Yes, the filter enum is a bitfield, so you have to construct a uint64 using something like pk_bitfield_from_enums() in C. You can do the same in go using `1 << enum_value` where the enum value is PK_FILTER_ENUM_UNKNOWN=0, PK_FILTER_ENUM_NONE=1, PK_FILTER_ENUM_INSTALLED=2 etc.

The NEWEST filter only returns the newest packages, so you don't get older packages showing up in the results.NOT_NEWEST only returns the packages that are not the newest installed and available packages.

**Comment by purpleidea** [comment]
@hughsie I would recommend adding some sort of method so that bitfield values can be specified to dbus by string, and not bit int, where it is error prone.

Secondly, I think I misunderstand how the NOT_NEWEST filter is supposed to work.

Cheers!

**Comment by james-lawrence** [comment]
I'm actually fine with the bit fields its relatively trivial to create them.

the problem really lies in the documentation for what each one means.

NOT_DEVEL, NOT_ARCH, etc are not clear about they are actually filtering on. What is the default state? if someone inserts 0 for the enum what does it mean for NOT_DEVEL to be 0? and DEVEL to be 0?
